# Wokflow  **TEST** con Full Bayesiana 

## Inicializacion

In [1]:
# limpio la memoria
Sys.time()
rm(list=ls(all.names=TRUE)) # remove all objects
gc(full=TRUE, verbose=FALSE) # garbage collection

[1] "2025-12-02 20:19:51 UTC"

used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  658631 35.2 1454648    77.7 1359152  72.6
Vcells 1222321  9.4 8388608    64.0 1975146  15.1

In [2]:
plocal <- list()

# 501
plocal$qcanaritos <- 5L
plocal$min_data_in_leaf <- 500L
plocal$learning_rate <- 1.0
plocal$gradient_bound <- 0.1
plocal$num_iterations <- 33

plocal$APO <- 1
plocal$ksemillerio <- 50


In [3]:
source("/home/vickydiliscia/dmeyf2025/src/churn/C3011/config_100271/_index.R")

In [4]:
source(file.path(PARAM$paths$functions, "index.R"))
source_functions(PARAM$paths$functions)

Sourcing: add_divisiones.R

Sourcing: add_ratio.R

Sourcing: agregar_clase_ternaria.R

Sourcing: agregar_var_rf.R

Sourcing: aplicar_dataquality_zeros.R

Sourcing: apply_sum_specs_multi.R

Sourcing: audit_nullear.R

Sourcing: build_variables_rotas.R

Sourcing: crear_proporcion.R

Sourcing: deflactar_por_ipc.R

Sourcing: features_historicos_optimizados.R

Loading required package: Rcpp

Sourcing: features_historicos.R

Sourcing: flip.R

Sourcing: interpolar_faltantes.R

Sourcing: lags.R

Sourcing: normalizar_ctrx_quarter.R

Sourcing: print_utils.R

Sourcing: run_features_historicas.R

Sourcing: tendencias_y_mas.R

Sourcing: ver_meses_rotos.R

✅ Todas las funciones fueron cargadas correctamente.



In [5]:
PARAM

$paths
$paths$experimento
[1] "APO3001-202107"

$paths$scriptfolder
[1] "C3001"

$paths$user
[1] "vickydiliscia"

$paths$base
[1] "/home/vickydiliscia/dmeyf2025"

$paths$bucket
[1] "/home/vickydiliscia/buckets/b1"

$paths$experiments
[1] "/home/vickydiliscia/buckets/b1/exp"

$paths$churn
[1] "/home/vickydiliscia/dmeyf2025/src/churn"

$paths$script
[1] "/home/vickydiliscia/dmeyf2025/src/churn/C3001"

$paths$datasets
[1] "/home/vickydiliscia/buckets/b1/datasets"

$paths$datasetcrudo2
[1] "/home/vickydiliscia/buckets/b1/datasets/competencia_02_crudo.csv.gz"

$paths$datasetcrudo3
[1] "/home/vickydiliscia/buckets/b1/datasets/competencia_03_crudo.csv.gz"

$paths$datasetclases
[1] "/home/vickydiliscia/buckets/b1/datasets/competencia_02.csv.gz"

$paths$functions
[1] "/home/vickydiliscia/dmeyf2025/src/churn/functions"

$paths$outputs
[1] "/home/vickydiliscia/buckets/b1/outputs"

$paths$results
[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107"

$paths$optimization
[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107/models"

$paths$prediction
[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107/prediction"

$paths$kaggle
[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107/kaggle"

$paths$backtesting
[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107/backtesting"


$dataquality
$dataquality$zeros
[1] "nullear"


$lgbm
$lgbm$boosting
[1] "gbdt"

$lgbm$objective
[1] "binary"

$lgbm$metric
[1] "custom"

$lgbm$first_metric_only
[1] FALSE

$lgbm$boost_from_average
[1] TRUE

$lgbm$feature_pre_filter
[1] FALSE

$lgbm$force_row_wise
[1] TRUE

$lgbm$verbosity
[1] -100

$lgbm$max_bin
[1] 31

$lgbm$num_iterations
[1] 9999

$lgbm$num_leaves
[1] 999

$lgbm$feature_fraction
[1] 0.5

$lgbm$seed
[1] 100271


$experiment
list()

$trainingstrategy
$trainingstrategy$testing
[1] 202107

$trainingstrategy$training
 [1] 201901 201902 201903 201904 201905 201906 201907 201908 201909 201910
[11] 201911 201912 202001 202002 202003 202004 202005 202007 202008 202009
[21] 202010 202011 202012 202101 202102 202103 202104 202105

$trainingstrategy$undersampling
[1] 0.05

$trainingstrategy$positivos
[1] "BAJA+1" "BAJA+2"


$train_final
$train_final$future
[1] 202107

$train_final$meses
 [1] 201901 201902 201903 201904 201905 201906 201907 201908 201909 201910
[11] 201911 201912 202001 202002 202003 202004 202005 202007 202008 202009
[21] 202010 202011 202012 202101 202102 202103 202104 202105

$train_final$training
 [1] 201901 201902 201903 201904 201905 201906 201907 201908 201909 201910
[11] 201911 201912 202001 202002 202003 202004 202005 202007 202008 202009
[21] 202010 202011 202012 202101 202102 202103 202104 202105

$train_final$undersampling
[1] 0.05


$train_entrega
$train_entrega$future
[1] 202109

$train_entrega$undersampling
[1] 0.05

$train_entrega$meses
 [1] 201901 201902 201903 201904 201905 201906 201907 201908 201909 201910
[11] 201911 201912 202001 202002 202003 202004 202005 202007 202008 202009
[21] 202010 202011 202012 202101 202102 202103 202104 202105 202106 202107

$train_entrega$training
 [1] 201901 201902 201903 201904 201905 201906 201907 201908 201909 201910
[11] 201911 201912 202001 202002 202003 202004 202005 202007 202008 202009
[21] 202010 202011 202012 202101 202102 202103 202104 202105 202106 202107


$features
$features$lags
[1] 2

$features$kmes
[1] TRUE

$features$normquarter
[1] TRUE


$user
[1] "vickydiliscia"

$experimento
[1] "APO3001-202107"

$folder
[1] "C3001"

$FE_rf
$FE_rf$arbolitos
[1] 20

$FE_rf$hojas_por_arbol
[1] 16

$FE_rf$datos_por_hoja
[1] 100

$FE_rf$mtry_ratio
[1] 0.2

$FE_rf$train
$FE_rf$train$training
[1] 202101 202102 202103


$FE_rf$lgb_param
$FE_rf$lgb_param$num_iterations
[1] 20

$FE_rf$lgb_param$num_leaves
[1] 16

$FE_rf$lgb_param$min_data_in_leaf
[1] 100

$FE_rf$lgb_param$feature_fraction_bynode
[1] 0.2

$FE_rf$lgb_param$boosting
[1] "rf"

$FE_rf$lgb_param$bagging_fraction
[1] 0.6321206

$FE_rf$lgb_param$bagging_freq
[1] 1

$FE_rf$lgb_param$feature_fraction
[1] 1

$FE_rf$lgb_param$max_bin

In [6]:
setwd("/content/buckets/b1/exp")
experimento_folder <- PARAM$experimento
dir.create(experimento_folder, showWarnings=FALSE)
setwd( paste0("/content/buckets/b1/exp/", experimento_folder ))

## Preprocesamiento

### Generacion de la clase_ternaria

In [7]:
paths <- c(PARAM$paths$datasetcrudo2,
           PARAM$paths$datasetcrudo3)

dataset <- rbindlist(
  lapply(paths, fread),
  use.names = TRUE,
  fill      = TRUE
)

dataset <- agregar_clase_ternaria(dataset)
setorder(dataset, foto_mes, clase_ternaria, numero_de_cliente)
dataset[, .N, .(foto_mes, clase_ternaria)]
Sys.time()

foto_mes clase_ternaria N     
1  201901   BAJA+1            645
2  201901   BAJA+2            729
3  201901   CONTINUA       122899
4  201902   BAJA+1            733
5  201902   BAJA+2            707
6  201902   CONTINUA       123961
7  201903   BAJA+1            708
8  201903   BAJA+2            751
9  201903   CONTINUA       124508
10 201904   BAJA+1            756
11 201904   BAJA+2            514
12 201904   CONTINUA       125268
13 201905   BAJA+1            517
14 201905   BAJA+2            692
15 201905   CONTINUA       125993
16 201906   BAJA+1            696
17 201906   BAJA+2            608
18 201906   CONTINUA       127430
19 201907   BAJA+1            611
20 201907   BAJA+2            680
21 201907   CONTINUA       128977
22 201908   BAJA+1            683
23 201908   BAJA+2            577
24 201908   CONTINUA       130883
25 201909   BAJA+1            581
26 201909   BAJA+2            582
27 201909   CONTINUA       132594
28 201910   BAJA+1            594
29 201910   BAJA+2            618
30 201910   CONTINUA       134798
⋮  ⋮        ⋮              ⋮     
67 202011   BAJA+1            456
68 202011   BAJA+2            616
69 202011   CONTINUA       159178
70 202012   BAJA+1            622
71 202012   BAJA+2            619
72 202012   CONTINUA       159756
73 202101   BAJA+1            622
74 202101   BAJA+2            825
75 202101   CONTINUA       160080
76 202102   BAJA+1            831
77 202102   BAJA+2           1032
78 202102   CONTINUA       160292
79 202103   BAJA+1           1039
80 202103   BAJA+2            951
81 202103   CONTINUA       161119
82 202104   BAJA+1            955
83 202104   BAJA+2           1130
84 202104   CONTINUA       161333
85 202105   BAJA+1           1134
86 202105   BAJA+2            842
87 202105   CONTINUA       161941
88 202106   BAJA+1            843
89 202106   BAJA+2           1134
90 202106   CONTINUA       162336
91 202107   BAJA+1           1137
92 202107   BAJA+2           1248
93 202107   CONTINUA       162211
94 202108   NA             163570
95 202108   BAJA+1           1252
96 202109   NA             165093

[1] "2025-12-02 20:21:05 UTC"

### Eliminacion de Features

In [8]:
# Salsa Magica para 202106
dataset[, mprestamos_personales := NULL ]
dataset[, cprestamos_personales := NULL ]
Sys.time()

[1] "2025-12-02 20:21:05 UTC"

### Data Quality

#### Flip Internet

In [9]:
flip(dataset, internet, month_to = 202009L)
Sys.time()

[1] "2025-12-02 20:21:05 UTC"

#### Tratamiento variables con meses rotos

In [10]:
variables_rotas <- build_variables_rotas(dataset)

if (PARAM$dataquality$zeros == "nullear") {
  dataset <- aplicar_dataquality_zeros(
    dataset,
    variables_rotas,
    modo = "nullear"
  )

} else if (PARAM$dataquality$zeros == "interpolar") {
  dataset <- aplicar_dataquality_zeros(
    dataset,
    variables_rotas,
    modo = "interpolar",
    treat_zero_as_na = TRUE,
    require_both     = TRUE,
    only_if_na       = TRUE,
    integer_prefix   = "^c",
    round_integers   = TRUE
  )

} else {
  message("No se aplicó tratamiento de dataquality.")
}

Sys.time()

[1] "2025-12-02 20:21:15 UTC"

In [11]:
if (PARAM$dataquality$zeros == "nullear") {
# Si usaste replace_all_in_month = FALSE (nullear solo ceros):
res <- audit_nullear(dataset, variables_rotas, month_col = "foto_mes",
                     replace_all_in_month = FALSE)

# Ver primeras filas y variables con problemas (si las hubiera)
res[ok_sin_ceros_marc | status == "ok"][]
}else{
    message("No se audita")
}
Sys.time()

variable                           tipo   
1  active_quarter                     integer
2  catm_trx                           integer
3  catm_trx_other                     integer
4  ccajas_consultas                   integer
5  ccajas_depositos                   integer
6  ccajas_extracciones                integer
7  ccajas_otras                       integer
8  ccajas_transacciones               integer
9  ccajeros_propios_descuentos        integer
10 ccallcenter_transacciones          integer
11 ccheques_depositados               integer
12 ccheques_depositados_rechazados    integer
13 ccheques_emitidos                  integer
14 ccheques_emitidos_rechazados       integer
15 ccomisiones_otras                  integer
16 cextraccion_autoservicio           integer
17 chomebanking_transacciones         integer
18 cmobile_app_trx                    integer
19 ctarjeta_debito_transacciones      integer
20 ctarjeta_master_descuentos         integer
21 ctarjeta_master_transacciones      integer
22 ctarjeta_visa_debitos_automaticos  integer
23 ctarjeta_visa_descuentos           integer
24 ctarjeta_visa_transacciones        integer
25 mactivos_margen                    numeric
26 matm                               numeric
27 matm_other                         numeric
28 mautoservicio                      numeric
29 mcajeros_propios_descuentos        numeric
30 mcheques_depositados               numeric
31 mcheques_depositados_rechazados    numeric
32 mcheques_emitidos                  numeric
33 mcheques_emitidos_rechazados       numeric
34 mcomisiones                        numeric
35 mcomisiones_otras                  numeric
36 mcuentas_saldo                     numeric
37 mextraccion_autoservicio           numeric
38 mpasivos_margen                    numeric
39 mrentabilidad                      numeric
40 mrentabilidad_annual               numeric
41 mtarjeta_master_consumo            numeric
42 mtarjeta_master_descuentos         numeric
43 mtarjeta_visa_consumo              numeric
44 mtarjeta_visa_descuentos           numeric
45 mttarjeta_visa_debitos_automaticos numeric
46 tcallcenter                        integer
47 thomebanking                       integer
48 tmobile_app                        integer
   meses_marcados                            filas_marcadas ceros_en_marcados
1  202006                                    153102         0                
2  202006                                    153102         0                
3  202006                                    153102         0                
4  202006                                    153102         0                
5  202006,202105                             317019         0                
6  202006                                    153102         0                
7  202006                                    153102         0                
8  202006                                    153102         0                
9  201910,202002,202006,202009,202010,202102 914901         0                
10 202006                                    153102         0                
11 202006                                    153102         0                
12 202006                                    153102         0                
13 202006                                    153102         0                
14 202006                                    153102         0                
15 201905,201910,202006                      416314         0                
16 202006                                    153102         0                
17 201910,202006                             289112         0                
18 202006                                    153102         0                
19 202006                                    153102         0                
20 201910,202002,202006,202009,202010,202102 914901         0                
21 202006                                    153102         0                
22 201904                        

[1] "2025-12-02 20:21:19 UTC"

## Data Drifting

In [13]:
library(data.table)

if (isTRUE(PARAM$FE_hist$Ipc$Run)) {
    ipc_dt <- fread(PARAM$FE_hist$specsipc)
    
    dataset <- deflactar_por_ipc(
        dt = dataset,
        ipc_dt = ipc_dt,
        replace = TRUE,
        base_mes = 202108L 
        )
    message(paste0("IPC aplicado con base ", 202109L ))
} else {
    message(paste0("Tecnica de data drifting no aplicada", 202109L ))
}
Sys.time()

IPC aplicado con base 202109



[1] "2025-12-02 20:21:29 UTC"

### Feature Engineering Intra-Mes

#### Variables Sugeridas

In [14]:
if (isTRUE(PARAM$FE_intrames$kmes)) {
  dataset[, kmes := foto_mes %% 100]
  message("Columna kmes creada")
} else {
  message("Columna kmes no creada")
}

if (isTRUE(PARAM$FE_intrames$normquarter)) {
  dataset <- normalizar_ctrx_quarter(dataset)
  message("Normalizacion ctrx_quarter aplicada")
} else {
  message("Normalizacion ctrx_quarter no aplicada")
}

if (isTRUE(PARAM$FE_intrames$ratio_payroll_edad)) {
    dataset[, mpayroll_sobre_edad := mpayroll / cliente_edad]  
    message("Columna mpayroll_sobre_edad creada")
} else {
    message("Columna mpayroll_sobre_edad creada")
}

Sys.time()

Columna kmes creada

Normalizacion ctrx_quarter aplicada

Columna mpayroll_sobre_edad creada



[1] "2025-12-02 20:21:34 UTC"

#### Sumas de Variables

In [15]:
spec_sumas <- fread(PARAM$FE_intrames$specsumas)

# Normalizar nombres de columnas
setnames(spec_sumas, tolower(names(spec_sumas)))
req <- c("new_var","inputs","active")
faltan <- setdiff(req, names(spec_sumas))
if (length(faltan)) stop("Faltan columnas en spec: ", paste(faltan, collapse=", "))

# Limpiar espacios en inputs y new_var
spec_sumas[, new_var := trimws(new_var)]
spec_sumas[, inputs  := trimws(inputs)]

# 2) Convertir 'active' a lógico robusto
to_bool <- function(x){
  if (is.logical(x)) return(x)
  x2 <- tolower(trimws(as.character(x)))
  x2 %in% c("true","t","1","yes","y","si","sí")
}
spec_sumas[, active := to_bool(active)]

dataset <- apply_sum_specs_multi(
  dataset,
  spec = spec_sumas,
  from_file = FALSE,
  sep_inputs = ","
)
Sys.time()

✅ 'mtarjeta_total_consumo' = suma_segura(mtarjeta_visa_consumo+mtarjeta_master_consumo)

✅ 'mtarjeta_total_debitos_automaticos' = suma_segura(mttarjeta_visa_debitos_automaticos+mttarjeta_master_debitos_automaticos)

✅ 'mfinanciacionlimite_total' = suma_segura(Master_mfinanciacion_limite+Visa_mfinanciacion_limite)

✅ 'msaldopesos_total' = suma_segura(Master_msaldopesos+Visa_msaldopesos)

✅ 'msaldodolares_total' = suma_segura(Master_msaldodolares+Visa_msaldodolares)

✅ 'mconsumopesos_total' = suma_segura(Master_mconsumospesos+Visa_mconsumospesos)

✅ 'mconsumodolares_total' = suma_segura(Master_mconsumosdolares+Visa_mconsumosdolares)

✅ 'mlimitecompra_total' = suma_segura(Master_mlimitecompra+Visa_mlimitecompra)

✅ 'madelantopesos_total' = suma_segura(Master_madelantopesos+Visa_madelantopesos)

✅ 'madelantodolares_total' = suma_segura(Master_madelantodolares+Visa_madelantodolares)

✅ 'mpagado_total' = suma_segura(Master_mpagado+Visa_mpagado)

✅ 'mpagopesos_total' = suma_segura(Master_mpag

[1] "2025-12-02 20:24:45 UTC"

### Feature Engineering Historico

#### Lags y Delta Lags

In [16]:
setorder(dataset, numero_de_cliente, foto_mes)

if (isTRUE(PARAM$FE_hist$Lags$Run)) {
  dataset <- crear_lags(dataset, lags = PARAM$FE_hist$Lags$Order)
} else {
  message("Lags no aplicados")
}

if (isTRUE(PARAM$FE_hist$DeltaLags$Run)) {
  dataset <- crear_delta_lags(dataset, lags = PARAM$FE_hist$DeltaLags$Order)
} else {
  message("Delta Lags no aplicados")
}

Sys.time()

Creando lags: 1, 2

✅ Lags creados: 350

Creando delta-lags: 1, 2

✅ Delta-lags creados: 350



[1] "2025-12-02 20:26:52 UTC"

#### Tendencias 6 meses

In [17]:
# 1) ¿Es data.table?
is.data.table(dataset)

# 2) ¿Qué columnas está intentando laguear?
id_col <- "numero_de_cliente"; time_col <- "foto_mes"
exclude <- c("clase_ternaria")

cols_lagueables <- setdiff(names(dataset), c(id_col, time_col, exclude))
length(cols_lagueables); head(cols_lagueables)

# 3) ¿Hay más de un registro por cliente (si no, los lags serían todo NA)?
dataset[, .N, by = id_col][N > 1L][1:5]

[1] TRUE

[1] 875

[1] "active_quarter"     "cliente_vip"        "internet"          
[4] "cliente_edad"       "cliente_antiguedad" "mrentabilidad"

numero_de_cliente N 
1 249221323         33
2 249227600         33
3 249227813         17
4 249231391         24
5 249234235         16

In [18]:

cols_lagueables <- intersect(cols_lagueables, colnames(dataset))
setorder(dataset, numero_de_cliente, foto_mes)

if( PARAM$FE_hist$Tendencias$run) {
    tendencias_y_mas(dataset,
    cols = cols_lagueables,
    ventana = PARAM$FE_hist$Tendencias$ventana, 
    tendencia = PARAM$FE_hist$Tendencias$tendencia,
    minimo = PARAM$FE_hist$Tendencias$minimo,
    maximo = PARAM$FE_hist$Tendencias$maximo,
    promedio = PARAM$FE_hist$Tendencias$promedio,
    ratioavg = PARAM$FE_hist$Tendencias$ratioavg,
    ratiomax = PARAM$FE_hist$Tendencias$ratiomax
  )
}

ncol(dataset)
Sys.time()

[1] 1180

[1] "2025-12-02 20:31:32 UTC"

#### Ratios, Medias moviles, Desvios y Streaks historicos

In [19]:
# Tomo specs de que variables crear
spec_hist <- fread(PARAM$FE_hist$spechistoricas) 

# Aplico variables
apply_rolling_with_kernel(dataset, spec_hist)

Sys.time()

Kernel Rcpp -> Var: mcuenta_corriente | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mcaja_ahorro | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mcuentas_saldo | win: 6 | ops: mean, sd, cv, ratioavg

Kernel Rcpp -> Var: mautoservicio | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mextraccion_autoservicio | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mtarjeta_visa_consumo | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mtarjeta_master_consumo | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mpagado_total | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mconsumototal_total | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: msaldototal_total | win: 6 | ops: mean, sd, cv, ratioavg

Kernel Rcpp -> Var: mpayroll_total | win: 6 | ops: mean, sd, cv, prop_gt0, ratioavg

Kernel Rcpp -> Var: mpagodeservicios | win: 6 | ops:

[1] "2025-12-02 20:32:24 UTC"

### Feature Engineering a partir de hojas de Random Forest

In [20]:
if( !require("lightgbm")) install.packages("lightgbm")
require("lightgbm")

Loading required package: lightgbm



In [21]:
# Agregar Variables de Random Forest

agregar_var_rf(PARAM)

Sys.time()

inicio AgregaVarRandomForest()
Fin construccion RandomForest
periodo =  201901 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201902 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201903 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201904 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201905 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201906 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201907 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7  8  9  10  11  12  13  14  15  16  17  18  19  20  
periodo =  201908 
Inicio prediccion
Fin prediccion
1  2  3  4  5  6  7

[1] "2025-12-02 20:36:19 UTC"

In [22]:
# guardar_db(dataset, "/home/vickydiliscia/datasets/dataset_APO_1")

## Modelado

No hay modelado, no se hace optimizacion de hiperparametros.

## Produccion

Las decisiones que se toman para la construccion del modelo final son:
* Los positvos son  POS={"BAJA+1", "BAJA+2"}, esta es una meticulosa decisión.
* Se entrena en los treinta meses del intervalo [201901, 202104]
* Se realiza undersampling al 5%
* Se utilizan los hiperparámetros optimos encontrados en la Bayesian Optimization
   * Se escala min_data_in_leaf

### Final Training Strategy

In [27]:
plocal

$qcanaritos
[1] 5

$min_data_in_leaf
[1] 500

$learning_rate
[1] 1

$gradient_bound
[1] 0.1

$num_iterations
[1] 33

$APO
[1] 5

$ksemillerio
[1] 50

$apo
[1] 1

In [28]:
# se filtran los meses donde se entrena el modelo final
dataset_train_final <- dataset[foto_mes %in% PARAM$train_final$training]

In [29]:
# canaritos
PARAM$train_final$lgbm$qcanaritos <- plocal$qcanaritos

cols0 <- copy(colnames(dataset_train_final))
filas <- nrow(dataset_train_final)

if( PARAM$train_final$lgbm$qcanaritos > 0 ) {
  for( i in seq(PARAM$train_final$lgbm$qcanaritos) ){
    dataset_train_final[, paste0("canarito_",i) := runif( filas) ]
  }

  # las columnas canaritos mandatoriamente van al comienzo del dataset
  cols_canaritos <- copy( setdiff( colnames(dataset_train_final), cols0 ) )
  setcolorder( dataset_train_final, c( cols_canaritos, cols0 ) )
}

Sys.time()

[1] "2025-12-02 20:38:18 UTC"

#### Registros cambio las proporciones de POS/NEG

In [30]:
# Undersampling, van todos los "BAJA+1" y "BAJA+2" y solo algunos "CONTINIA"

set.seed(PARAM$semilla_primigenia, kind = "L'Ecuyer-CMRG")
dataset_train_final[, azar := runif(nrow(dataset_train_final))]
dataset_train_final[, training := 0L]

dataset_train_final[
  (azar <= PARAM$train_final$undersampling | clase_ternaria %in% c("BAJA+1", "BAJA+2")),
  training := 1L
]

dataset_train_final[, azar:= NULL] # elimino la columna azar

### Target Engineering

In [31]:
# paso la clase a binaria que tome valores {0,1}  enteros
#  BAJA+1 y BAJA+2  son  1,   CONTINUA es 0
#  a partir de ahora ya NO puedo cortar  por prob(BAJA+2) > 1/40

dataset_train_final[,
  clase01 := ifelse(clase_ternaria %in% c("BAJA+2","BAJA+1"), 1L, 0L)
]

### Final Models

Aqui SIEMPRE voy a hacer un semillerio, independientemente de si en la Bayesian Optimization calculé un semillerio en cada iteración.
<br> Entreno un LightGBM para cada semilla,  y guardo el modelo dentro de la carpeta  **modelitos**
<br> Intencionalmente en una primera etapá se generan los modelos y graban, y en una segunda etapa se leen eso modelos y se aplican a los datos del futuro

APO controla cuantas veces se repite el modelo, que se usa para promediar ganancias y reportar en la Pseudo Competencia algo razonable
<br> El modelo puede ser un LightGBM simple (ksemillerio==1)  o un Ensemble Semillerio( ksemillerio > 1 )
<br> Lamentablmente APO necesita utilizar muchas semillas, y eso demanda TIEMPO de corrida

In [32]:

PARAM$train_final$lgbm$param_completo <-  list(
  boosting= "gbdt",
  objective= "binary",
  metric= "custom",
  first_metric_only= FALSE,
  boost_from_average= TRUE,
  feature_pre_filter= FALSE,
  force_row_wise= TRUE,
  verbosity= -100,

  seed= PARAM$semilla_primigenia,

  max_bin= 31L,
  min_data_in_leaf= plocal$min_data_in_leaf,  #este ya es el valor default de LightGBM

  num_iterations= plocal$num_iterations, # dejo libre la cantidad de arboles, zLightGBM se detiene solo
  num_leaves= 9999L, # dejo libre la cantidad de hojas, zLightGBM sabe cuando no hacer un split
  learning_rate= plocal$learning_rate,  # se lo deja en 1.0 para que si el score esta por debajo de gradient_bound no se lo escale
    
  feature_fraction= 0.50, # un valor equilibrado, habra que probar alternativas ...
    
  canaritos= PARAM$train_final$lgbm$qcanaritos, # fundamental en zLightGBM, aqui esta el control del overfitting
  gradient_bound= plocal$gradient_bound   # default de zLightGBM
)

Sys.time()

[1] "2025-12-02 20:38:29 UTC"

In [34]:
# Semillerio Final
PARAM$train_final$APO <- plocal$APO
PARAM$train_final$ksemillerio  <- plocal$ksemillerio

PARAM$train_final$cortes <- c(9000, 9500, 10000, 10500, 11000, 11500, 12000, 12500, 13000)

In [35]:
detach("package:lightgbm", unload= TRUE)

In [36]:
if( !require("zlightgbm") ) install.packages("https://storage.googleapis.com/open-courses/dmeyf2025-e4a2/zlightgbm_4.6.0.99.tar.gz", repos= NULL, type= "source")
require("zlightgbm")

Loading required package: zlightgbm

Registered S3 methods overwritten by 'zlightgbm':
  method                 from    
  dimnames<-.lgb.Dataset lightgbm
  dim.lgb.Dataset        lightgbm
  dimnames.lgb.Dataset   lightgbm
  predict.lgb.Booster    lightgbm
  print.lgb.Booster      lightgbm
  summary.lgb.Booster    lightgbm



In [37]:
if(!require("primes")) install.packages("primes")
require("primes")

Loading required package: primes



In [38]:
primos <- generate_primes(min = 100000, max = 1000000)
set.seed(PARAM$semilla_primigenia, kind = "L'Ecuyer-CMRG")
PARAM$train_final$semillas <- sample(primos)[seq( PARAM$train_final$APO*PARAM$train_final$ksemillerio )]
PARAM$train_final$semillas

[1] 827009 909409 436061 917729 237203 258283 156329 604759 527897 504967
 [11] 840941 906329 304631 489871 193751 728899 897067 843907 360197 723553
 [21] 972557 937991 623983 529517 234473 398509 237619 674789 909599 159773
 [31] 151531 703249 442769 534553 257171 613451 914813 246497 680353 519119
 [41] 233419 362233 752251 141677 302221 400277 450361 253439 206651 952481
 [51] 279689 343709 167801 255181 691949 363329 365611 296557 453799 313949
 [61] 955333 649427 272651 167269 897499 608897 540697 354551 498409 597637
 [71] 518311 314549 783571 998399 265567 995881 740161 398917 294439 561521
 [81] 904627 997609 858623 932207 478651 887827 477863 392213 581551 545873
 [91] 294461 161387 711913 242273 405947 456607 364937 826193 251171 270191
[101] 957889 847577 109741 806261 965533 156589 635603 148123 131759 143113
[111] 329587 227593 123407 848051 644671 991961 102337 634759 754381 208469
[121] 946037 490271 517577 320081 423769 828731 559513 925949 518417 420097
[131] 382241 214069 208513 890737 827231 788891 705013 475907 255877 982777
[141] 533581 146681 125453 343631 963097 383767 430343 766873 406247 550073
[151] 997147 997553 264527 799891 507809 183661 739111 870917 396637 518809
[161] 964583 599477 613637 820399 890317 390959 403567 136397 261577 153997
[171] 932567 503249 207433 254827 516839 119657 569581 906371 146743 819061
[181] 403993 209159 478343 575203 182657 392593 912343 704087 142537 932749
[191] 415673 465299 453877 904873 150989 155699 499621 183877 996173 990503
[201] 112759 873553 344269 838441 351157 846061 362459 646897 703837 873091
[211] 492431 168409 754073 395321 924499 815209 890761 920833 652727 367453
[221] 342059 737119 468887 169483 757727 516907 527441 819271 301141 358229
[231] 502819 668609 319973 325849 969863 542599 834941 335633 422209 829643
[241] 794779 354377 438941 141283 113489 509573 478831 161659 369029 632231

In [39]:
campos_buenos <- setdiff(
  colnames(dataset_train_final),
  c( "clase_ternaria", "clase01", "training", "azar")
)

In [40]:
# dejo los datos en formato LightGBM
dtrain_final <- lgb.Dataset(
  data= data.matrix(dataset_train_final[training == 1L, campos_buenos, with= FALSE]),
  label= dataset_train_final[training == 1L, clase01],
  free_raw_data= FALSE
)

cat("filas", nrow(dtrain_final), "columnas", ncol(dtrain_final), "\n")
Sys.time()

filas 237459 columnas 1505 


[1] "2025-12-02 20:39:13 UTC"

In [41]:
getwd()

[1] "/home/vickydiliscia/buckets/b1/exp/APO3001-202107"

In [42]:
# genero los modelitos
dir.create( "modelitos", showWarnings= FALSE)

param_completo <- copy( PARAM$train_final$lgbm$param_completo)

for( sem in PARAM$train_final$semillas ) {

  arch_modelo <- paste0("./modelitos/mod_", sem, ".txt")
  if( !file.exists( arch_modelo ) )
  {
    param_completo$seed <- sem

    modelito <- lgb.train(
      data= dtrain_final,
      param= param_completo
    )

    lgb.save( modelito, filename= arch_modelo)
    rm(modelito)
    gc()
  }
}

Sys.time()

[1] "2025-12-02 21:08:16 UTC"

### Scoring

Se hace el predict() del modelo en los datos del futuro

In [43]:
dfuture <- dataset[foto_mes %in% PARAM$train_final$future ]

cols0 <- copy(colnames(dfuture))
filas <- nrow(dfuture)

if( PARAM$train_final$lgbm$qcanaritos > 0 ) {
  for( i in seq(PARAM$train_final$lgbm$qcanaritos) ){
    dfuture[, paste0("canarito_",i) := runif( filas) ]
  }

  # las columnas canaritos mandatoriamente van al comienzo del dataset
  cols_canaritos <- copy( setdiff( colnames(dfuture), cols0 ) )
  setcolorder( dfuture, c( cols_canaritos, cols0 ) )
}

In [44]:
# dataset de future, donde en este caso estoy haciendo testing

mfuture <- data.matrix(dfuture[, campos_buenos, with= FALSE])

dfuture[, ganancia := ifelse(clase_ternaria=="BAJA+2", 780000, -20000)]

In [45]:
mganancias <- matrix( nrow=PARAM$train_final$APO, ncol= length(PARAM$train_final$cortes) )

if( file.exists("prediccion.txt") )
  file.remove("prediccion.txt")

In [ ]:
# aplico el modelo a los datos del future
dir.create("predicciones_modelitos", showWarnings = FALSE)

for( vapo in seq(PARAM$train_final$APO) ) {
  # inicializacion en CERO
  vpred_acum <- rep(0.0, nrow(dfuture))
  qacumulados <- 0

  desde <- 1 + (vapo-1)*PARAM$train_final$ksemillerio
  hasta <- desde + PARAM$train_final$ksemillerio - 1
  semillas <- PARAM$train_final$semillas[desde:hasta]

  for( sem in semillas ) {

    arch_modelo <- paste0("./modelitos/mod_", sem, ".txt")
    if( file.exists( arch_modelo ) )
    {
      modelo_final <- lgb.load(arch_modelo) # leo del disco
      # hago el predict() y acumulo
        
      vpred <- predict(modelo_final, mfuture)   # <- guardo primero la prediccion individual
      fwrite(
        data.table(
        numero_de_cliente = dfuture$numero_de_cliente,
        prob              = vpred
      ), file = sprintf("predicciones_modelitos/prediccion_semilla_%s.txt", sem)
      )
        
      # ---- luego acumulo al ensemble ----
      vpred_acum <- vpred_acum + vpred
      qacumulados <- qacumulados + 1

      rm(modelo_final)
      gc()
    }
  }

  if( qacumulados > 0 ) {
    vpred_acum <- vpred_acum / qacumulados  # paso a probabildiad
    # tabla de prediccion, puede ser util para futuros ensembles
    #  ya que le modelo ganador va a ser un ensemble de LightGBMs

    tb_prediccion <- dfuture[, list(numero_de_cliente, foto_mes, ganancia)]
    tb_prediccion[, meta_modelo := vapo]
    tb_prediccion[, prob := vpred_acum ]
    setorder( tb_prediccion, -prob )
    tb_prediccion[, gan_acum := cumsum(ganancia)]
    tb_prediccion[, ganancia := NULL ]

    # acumulo las ganancias
    for( icor in seq(length(PARAM$train_final$cortes)) ){
      mganancias[ vapo, icor ] <- tb_prediccion[ PARAM$train_final$cortes[icor], gan_acum ]
    }

    # grabo las probabilidades del modelo
    fwrite(tb_prediccion,
      file= "prediccion.txt",
      sep= "\t",
      append= TRUE
    )

    rm(tb_prediccion)
    gc()
  }
}

Sys.time()

In [ ]:
mganancias

### Clasificacion

Se tomó la decisión de enviar a los 11000 registros con mayor probabilidad de POS={"BAJA+1","BAJA+"}
<br> esto se determinó en forma artesanal analizando meses anterior
<br> esta es una muy importante decisión 

In [ ]:
# genero archivos con los  "envios" mejores
dir.create("kaggle", showWarnings=FALSE)

tb_prediccion <- fread("prediccion.txt")

In [ ]:
# genero archivos de fantasia, que NO son el que voy a subir a la Pseudo Competencia Kaggle
envios <- 11000

for( vapo in seq(PARAM$train_final$APO) ) {
  if( tb_prediccion[meta_modelo==vapo, .N] > 0 ) {
    tb_pred <- tb_prediccion[meta_modelo==vapo]
    setorder( tb_pred, -prob )
    tb_pred[, Predicted := 0L] # seteo inicial a 0
    tb_pred[1:envios, Predicted := 1L] # marco los primeros

    archivo_kaggle <- paste0("./kaggle/KA", PARAM$experimento, "_", vapo, "_", envios, ".csv")

    # grabo el archivo
    fwrite(tb_pred[, list(numero_de_cliente, Predicted)],
      file= archivo_kaggle,
      sep= ","
    )

    rm( tb_pred )
    gc()
  }
}

Sys.time()

### Subida a Pseudo Kaggle

Aqui viene la verdadera magia de  APO = A Prueba Overfiteres
<br>Se sube un submit a la Pseudo Competencia Kaggle cuya ganancia coincide casi exactamente con la ganancia MEDIA , promediada  APO veces

In [ ]:
colmedias <- colMeans( mganancias, na.rm=TRUE )
mcorte_mejor <- max(colmedias, na.rm=TRUE)
icorte_mejor <- which.max( colmedias )
corte_mejor <- PARAM$train_final$cortes[icorte_mejor]

In [ ]:
tbl <- as.data.table( as.list( colmedias ) )
colnames(tbl) <- paste0( "e", PARAM$train_final$cortes )
tbl[, experimento := PARAM$experimento ]

exp_gral <- "/content/buckets/b1/exp/apo-gral"
dir.create(exp_gral, showWarnings=FALSE)
fwrite( tbl,
  file= paste0( exp_gral, "/tb_experimentos.txt"),
  sep= "\t",
  append= TRUE
)

In [ ]:
colnames( mganancias ) <- paste0( "e", PARAM$train_final$cortes )
tbl_local <- as.data.table( mganancias )

fwrite( tbl_local,
  file= "tb_apo.txt",
  sep= "\t"
)

In [ ]:
icerca <- which.min(  abs( tb_prediccion$gan_acum - mcorte_mejor ) )
vmodelo <- tb_prediccion[ icerca, meta_modelo ]
tb_pred <- tb_prediccion[meta_modelo==vmodelo]

mcorte_mejor
icerca
tb_prediccion[ icerca]

In [ ]:
icerca <- which.min(  abs( tb_pred$gan_acum - mcorte_mejor ) )
icerca

In [ ]:
icerca <- which.min(  abs( tb_prediccion$gan_acum - mcorte_mejor ) )
vmodelo <- tb_prediccion[ icerca, meta_modelo ]
tb_pred <- tb_prediccion[meta_modelo==vmodelo]

icerca <- which.min(  abs( tb_pred$gan_acum - mcorte_mejor ) )
tb_pred[, Predicted := 0L] # seteo inicial a 0
tb_pred[1:icerca, Predicted := 1L] # marco los primeros

archivo_pseudo_kaggle <- paste0("./kaggle/KA", PARAM$experimento, "_",  icerca, ".csv")

# grabo el archivo
fwrite(tb_pred[, list(numero_de_cliente, Predicted)],
  file= archivo_pseudo_kaggle,
  sep= ","
)


In [ ]:
# la subida a Kaggle
comando <- "kaggle competitions submit"
competencia <- "-c  test-202106"
arch <- paste( "-f", archivo_pseudo_kaggle)
mensaje <-  paste0( "-m 'exp=", PARAM$experimento,
  "  ", paste(names(plocal), plocal, sep= "=", collapse= ";" ),
  " envios=", icorte_mejor,"'")

                    
linea <- paste( comando, competencia, arch, mensaje)
salida <- system(linea, intern=TRUE)
cat(salida)

In [ ]:
Sys.time()

# Entrega Final Final

In [ ]:
PARAM$train_entrega

In [ ]:
# se filtran los meses donde se entrena el modelo final
dataset_train_final <- dataset[foto_mes %in% PARAM$train_entrega$meses]

In [ ]:
# Undersampling, van todos los "BAJA+1" y "BAJA+2" y solo algunos "CONTINIA"

set.seed(PARAM$semilla_primigenia, kind = "L'Ecuyer-CMRG")
dataset_train_final[, azar := runif(nrow(dataset_train_final))]
dataset_train_final[, training := 0L]

dataset_train_final[
  (azar <= PARAM$train_entrega$undersampling | clase_ternaria %in% c("BAJA+1", "BAJA+2")),
  training := 1L
]

dataset_train_final[, azar:= NULL] # elimino la columna azar

In [ ]:
dataset_train_final[,
  clase01 := ifelse(clase_ternaria %in% c("BAJA+2","BAJA+1"), 1L, 0L)
]

In [ ]:


cols0 <- copy(colnames(dataset_train_final))
filas <- nrow(dataset_train_final)

for( i in seq(plocal$qcanaritos) ){
  dataset_train_final[, paste0("canarito_",i) := runif( filas) ]
}

# las columnas canaritos mandatoriamente van al comienzo del dataset
cols_canaritos <- copy( setdiff( colnames(dataset_train_final), cols0 ) )
setcolorder( dataset_train_final, c( cols_canaritos, cols0 ) )

Sys.time()

In [ ]:
plocal <- list()

# 501
plocal$qcanaritos <- 5L
plocal$min_data_in_leaf <- 20L
plocal$learning_rate <- 1.0
plocal$gradient_bound <- 0.1

plocal$APO <- 5
plocal$ksemillerio <- 1

In [ ]:
# los campos que se van a utilizar

campos_buenos <- setdiff(
  colnames(dataset_train_final),
  c("clase_ternaria", "clase01", "training")
)

In [ ]:
dtrain_final <- lgb.Dataset(
  data= data.matrix(dataset_train_final[training == 1L, campos_buenos, with= FALSE]),
  label= dataset_train_final[training == 1L, clase01],
  free_raw_data= FALSE
)

cat("filas", nrow(dtrain_final), "columnas", ncol(dtrain_final), "\n")
Sys.time()

In [ ]:
# entreno el modelo

modelo_final <- lgb.train(
  data= dtrain_final,
  param= PARAM$lgbm
)

Sys.time()

In [ ]:
# entreno el modelo

modelo_final <- lgb.train(
  data= dtrain_final,
  param= param_completo
)

Sys.time()

In [ ]:
# genero los modelitos
dir.create( "modelitos_produccion", showWarnings= FALSE)


for( sem in PARAM$train_final$semillas ) {

  arch_modelo <- paste0("./modelitos_produccion/mod_", sem, ".txt")
  if( !file.exists( arch_modelo ) )
  {
    param_completo$seed <- sem

    modelito <- lgb.train(
      data= dtrain_final,
      param= param_completo
    )

    lgb.save( modelito, filename= arch_modelo)
    rm(modelito)
    gc()
  }
}

Sys.time()

In [ ]:
# 1) Datos del future (entrega)
dfuture <- dataset[ foto_mes %in% PARAM$train_entrega$future ]

filas <- nrow(dfuture)

# 2) Agrego canaritos (solo una vez)
for( i in seq(plocal$qcanaritos) ){
  dfuture[, paste0("canarito_", i) := runif(filas) ]
}

# 3) Matriz de features para predict (sin clase)
mfuture <- data.matrix(dfuture[, campos_buenos, with = FALSE])

# 4) Carpeta donde voy a guardar las predicciones por semilla
dir.create("predicciones_produccion", showWarnings = FALSE)

# 5) Loop por semillas: cargo modelo, predigo y guardo
for( sem in PARAM$train_final$semillas ) {

  arch_modelo <- paste0("./modelitos_produccion/mod_", sem, ".txt")

  if( file.exists(arch_modelo) ) {

    modelo_final <- lgb.load(arch_modelo)

    prediccion <- predict(
      modelo_final,
      mfuture
    )

    # guardo SOLO numero_de_cliente y prob
    fwrite(
      data.table(
        numero_de_cliente = dfuture$numero_de_cliente,
        prob              = prediccion
      ),
      file = sprintf("predicciones_produccion/pred_semilla_%s.csv", sem)
    )

    rm(modelo_final, prediccion)
    gc()
  }
}
